File: prep/implementation/practice_iterations/main_1/PermutationWithRepetitions.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class PermutationWithRepetitions {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var perms = permute(a);
        for (var perm : perms) {
          for (int p : perm) {
            System.out.printf("%d ", p);
          }
          System.out.println();
        }
      }
    }
  }

  // O(n x n!) / O(n x n!)
  private static List<List<Integer>> permute(int[] a) {
    Arrays.sort(a);

    List<List<Integer>> perms = new ArrayList<>();
    Set<Integer> visited = new HashSet<>();

    permute(a, 0, visited, new ArrayList<>(), perms);

    return perms;
  }

  private static void permute(int[] a, int currIdx, Set<Integer> visited,
                              List<Integer> currPerm,
                              List<List<Integer>> allPerms) {

    if (currPerm.size() == a.length) {
      allPerms.add(new ArrayList<>(currPerm));
      return;
    }

    if (currIdx == a.length) {
      return;
    }

    for (int i = 0; i < a.length; i++) {
      if (visited.contains(i)) {
        continue;
      }

      if (i > 0 && (a[i] == a[i - 1] && visited.contains(i - 1))) {
        continue;
      }

      visited.add(i);
      currPerm.add(a[i]);
      permute(a, currIdx + 1, visited, currPerm, allPerms);
      visited.remove(i);
      currPerm.remove(currPerm.size() - 1);
      permute(a, currIdx + 1, visited, currPerm, allPerms);
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/tidbits_2/permutation_with_repetitions.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

void permute(const vector<int> &a, int curr_idx, int n, vector<bool> &visited,
             vector<int> &curr_perm, vector<vector<int>> &perms) {
  if (curr_perm.size() == n) {
    perms.push_back(curr_perm);
    return;
  }

  if (curr_idx == n) {
    return;
  }

  for (int i = 0; i < n; i++) {
    if (visited[i]) {
      continue;
    }

    if (i > 0 && (a[i] == a[i - 1]) && visited[i - 1]) {
      continue;
    }

    visited[i] = true;
    curr_perm.push_back(a[i]);
    permute(a, curr_idx + 1, n, visited, curr_perm, perms);
    visited[i] = false;
    curr_perm.pop_back();
    permute(a, curr_idx + 1, n, visited, curr_perm, perms);
  }
}

vector<vector<int>> permute(vector<int> &a) {
  sort(a.begin(), a.end());

  vector<vector<int>> perms;
  vector<int> curr_perm;
  vector<bool> visited(a.size(), false);

  permute(a, 0, a.size(), visited, curr_perm, perms);
  return perms;
}

// O(n! * n) / O(n! * n)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto perms = permute(a);
    for (auto perm : perms) {
      for (int p : perm) {
        cout << p << " ";
      }
      cout << "\n";
    }
  }

  return 0;
}